Add note to Matrix.Multiply methods to not use a parameter as both ref and out

Basewq opened the pull request:
EDIT: This pull has been changed to note the quirk with `Matrix.Multiply` rather than change it.
# PR Details

<!--- Provide a general summary of your changes in the Title above -->
Add documentation to not use a `ref` variable as the `out` parameter for `Matrix.Multiply`, ie.
```
var matrix1 = // Some matrix
var matrix2 = // Some other matrix
Matrix.Multiply(ref matrix1, ref matrix2, out matrix2); // This will not correctly calculate matrix1 * matrix2
```

## Description

<!--- Describe your changes in detail -->
Add a `<remarks>` tag to `Matrix.Multiply`.
Fix code where `Matrix.Multiply` was used in the way above.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
We cannot fix `Matrix.Multiply` to calculate the new matrix defensively because of performance degradation, so we tell them not to do it.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [X] Docs change / refactoring / dependency upgrade
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Kryptos-FR commented:
Unless I am missing something, this PR does the exact opposite of what it claims to fix.

In other words, it always does a copy where in some case we could prevent it.

Consider:

    Matrix m1 = ...;
    Matrix m2 = ...;
    Matrix m3;

    Matrix.Multiply(ref m1, ref m2, out m3);

    // use m3...

    matrix1...  // change something

    // No copy in previous code, unnecessary copy in new code
    Matrix.Multiply(ref m1, ref m2, out m3);

What we should do instead is document that matrix3 must be a different reference from matrix1 or matrix2. But adding a temporary variable causes a copy in all cases.

What you are proposing is equivalent to (without change):

    Matrix m1, m2, m3;
    Matrix.Multiply(ref m1, ref m2, out m3);
    m2 = m3;

I would prefer the copy to be explicit at the calling site instead of hidden in the method.

Basewq commented:
I might be a little confused on your example.
As it stands having 3 separate Matrix variables works, regardless of this fix or not.
Having only 2 Matrix, where one is also the output, currently does not work because it is partially writing over itself then the overwritten fields are then being used the in subsequent calculations.

Yes, what I'm proposing is nearly equivalent to your bottom example, but technically `m3` will not exist.

No, matrix3 should not be documented to be different that `matrix1` & `matrix2`.

A real simple example (this is from Xenko's source code) is changing the code from:
`physicsTransform = Matrix.Multiply(ColliderShape.PositiveCenterMatrix, physicsTransform);`
to
`Matrix.Multiply(ref ColliderShape.PositiveCenterMatrix, ref physicsTransform, out physicsTransform);`

However one primary example of using it is for chaining your calculations:
```
Matrix.Multiply(ref positionMatrix, ref translationMatrix, out var finalTransformMatrix);
Matrix.Multiply(ref finalTransformMatrix, ref scaleMatrix, out finalTransformMatrix);
Matrix.Multiply(ref finalTransformMatrix, ref rotationMatrix, out finalTransformMatrix);
```
(or whatever the correct order of matrix operations).


As a side note, XNA/Monogame supports doing this. The whole purpose of these `ref` versions were because copying large structs around performed badly and people wanted a faster version.
Yes, the code may look stupid and uglier but that's a limitation of the C# language.

Kryptos-FR commented:
What I meant is that your change doesn't remove the copy, it just hides it. There is a copy happening inside the method now : the temp variable.

But it does add a copy for other code paths that did not need chaining calls.

Creating worse performance for convenience is not a good trade off in my opinion.

Basewq commented:
Ok, I understand. Yes, it's essentially creating a defensive copy in that sense.
And yes, it degrades the performance slightly because of this.

I would say that the results of these should all be equivalent, but they are not:
```
m1 *= m2;
m1 = Matrix.Multiply(m1, m2);
Matrix.Multiply(ref m1, ref m2, out m1);
```
This goes against the behavior of every other method that follow similar signatures.

If you still disagree, feel free to reject this.

Basewq commented:
> 
> > No, matrix3 should not be documented to be different that matrix1 & matrix2.
> 
> Yes it should. Passing the same variable to both a ref and out argument is considered bad practice and a code smell.

I don't consider it bad practice. It reduces code noise by not requiring a dummy variable while still using the `ref` feature.
FYI, the code base has a few of these already in place, and I find it perfectly readable, eg.

https://github.com/xenko3d/xenko/blob/master/sources/engine/Xenko.Graphics/SpriteBatch.cs#L506

https://github.com/xenko3d/xenko/blob/master/sources/engine/Xenko.Graphics/SpriteFont.cs#L236-L237

https://github.com/xenko3d/xenko/blob/master/sources/engine/Xenko.Graphics/Sprite3DBatch.cs#L170-L175

https://github.com/xenko3d/xenko/blob/28245aff55cea3ceacaef2ea3902657cd06f81b6/sources/engine/Xenko.Engine/Engine/EntityTransformExtensions.cs#L157-L160

Amusingly enough, I discover there is one single code that used `Matrix.Multiply` in the way you'd think it should work, but it turns out it doesn't:

https://github.com/xenko3d/xenko/blob/080774745151b858794111ac128a9a473b517d19/sources/engine/Xenko.Engine/Engine/TransformComponent.cs#L312

In the end, no hard feelings if this shouldn't be merged.

tebjan commented:
How about adding this as a new operation? something like `Matrix.MultiplyInPlace` with a proper help text what the difference to the other is.

Basewq commented:
Personally, I think the safer version should be the default version to be consistent with how the other `ref` -> `out`, and this also makes any XNA/Monogame developers not get caught out if they ported their code.
However, I understand if this this not to be done, since the majority of `Matrix.Multiply` used in the current code base would have to change to the unsafer version.

As a note, there's actually **three** versions, with the *exact* same code underneath:
```
public static void MultiplyTo(ref Matrix left, ref Matrix right, out Matrix result)
public static void Multiply(ref Matrix left, ref Matrix right, out Matrix result)
public static void MultiplyRef(ref Matrix left, ref Matrix right, ref Matrix result)
```

The history log shows nothing, but I suspect either `MultiplyTo` or `Multiply` used to be the safe version, but a dev didn't notice the subtle difference and thought both could be the same.

I don't understand why `MultiplyRef` exists, maybe the person didn't think `ref` & `out` were the same from an output perspective?

Basewq commented:
I did some (poor) benchmarking to see if there's any alternative to defensive copying.
It seems doing it like Monogame where you calculate each field as a free variable, then assign them all back to the out struct was faster than a temp variable, however it still performs badly against the current version.

I tested looping through an array with 1,000,000 items and multiplying once, and the current version beat the free variable assignment version by ~15%.
Testing it with chaining multiply matrices (which seems to prevents a lot of the optimization?), the difference was closer to 3%.

Anyway, I've changed this pull to just add a note to not do the inline trick. It probably deserves to be in the summary tag so it'll show up in intellisense, since it's behavior deviates from the rest, but for the time it's just left in the remarks tag.

Also fixed TransformComponent and added a note to ensure no one tries to inline it.

Looking through the code base, there's a few places which will benefit from changing their matrix multiplication from non-ref to ref version, but it'll need to declare a dummy variable. More code clutter, but there should still be some performance boost.

tebjan commented on sources/engine/Xenko.Engine/Engine/TransformComponent.cs:
```diff
@@ -309,7 +309,8 @@ internal void UpdateWorldMatrixInternal(bool recursive)
                     }
 
-                    Matrix.Multiply(ref WorldMatrix, ref scene.WorldMatrix, out WorldMatrix);
+                    var curWorldMatrix = WorldMatrix;   // Must make a copy to use as the ref parameter, otherwise matrix will not be calculated correctly
```
wow, just out of interest, was this a bug all along? do you remember how it affected the scene?
I had some transformation bugs and wonder whether it came from this.

Basewq replied:
Surprisingly, this bug doesn't manifest in game.
I haven't rigorously tested this, but I tested a few matrix multiplication, and simple translations & scaling as the second matrix still multiply correctly.
Once you have rotation, it'll calculation incorrectly.
Looking at Game Studio, it seems you can only do translation offsets to scenes, so you can't encounter this bug through this code (as far as I can tell).

Eideren replied:
We could perhaps assert/check in debug mode for cases like these, fix and compare pointers of refs and outs, if they are the same throw ?

tebjan replied:
that's a pretty good idea!

Basewq replied:
I am willing to make a commit for this (which can be accepted or rejected), though this is a heavily used method, so it might have an impact a user's debug build.
As far as I can tell, it needs to be written like this?
```
unsafe
{
    fixed (void* leftPtr = &left)
    fixed (void* rightPtr = &right)
    fixed (void* resultPtr = &result)
    {
        System.Diagnostics.Debug.Assert(leftPtr != resultPtr, "This method does not support 'left' parameter being the same as 'result' parameter.");
        System.Diagnostics.Debug.Assert(rightPtr != resultPtr, "This method does not support 'right' parameter being the same as 'result' parameter.");
    }
}
```

How do I expose this to propagate a user's build?
I think `#if DEBUG` doesn't work for user consumption?
Do I wrap it in another method but tag it `[Conditional("DEBUG")]`?

Eideren replied:
The perf hit in debug mode is around 33% of the base cost of the function regardless of if we're using an exception or an assert and there's no elegant way to conditionally run logic based on the entry assembly's build type; it can't work properly with users debug build if we want optimal perfs in release so I don't think it's worthwhile to have this check in place in the end, sorry about that.

tebjan replied:
@Eideren ok, that's a lot for a method that get's used all over the place.
@Basewq is the comment now in the summary or remarks? since it's important info, the programmer should see it in IntelliSense, i'd say.

tebjan replied:
> it can't work properly with users debug build if we want optimal perfs

is `#ifdef DEBUG` not working here?

Eideren replied:
Preprocessors are taken care of in the compilation process, the JIT doesn't take care of this kind of things so release dlls are already stripped out of any ``#debug``. I don't think that we can provide both debug and release versions within nugget packages and conditionally include one or the other depending on the application's build mode. 

Users would have to explicitly take the debug dlls to ever have these kinds of checks.

You can detect whether the entry assembly runs in debug or release mode but that's kind of a hack since we don't know whether the user intended the engine to run in debug mode based solely on the entry assembly's build mode or if he just has a weird project structure. 
We would also have to pay the price of having at least one branch in release mode, I don't think the issue here warrants it.
Perhaps once c# has function pointers we could take a look at this again.

tebjan replied:
yes, that would require to ship the debug dlls... true. not important enough. but I would move the info into the summary. @Basewq

Basewq replied:
Ok, promoted the note to summary tag.

Regarding the debug note, I've been reading up on the `Conditional` attribute, and I think this allows for the kind of branching you're looking for.
[eg. this example show](https://www.dotnetperls.com/conditional) methods are always included in every dll, but only runs when certain symbols are declared. `Debug.Write` is probably the best example.
[The IL do not get emitted](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.conditionalattribute?view=netframework-4.8) if the symbol isn't declared.
We could technically define a new symbol for users to opt in to get this extra validation, however looking at the code base, there isn't anything like this for public use already so this would be a completely new feature introduced (for advanced users...).
That said, let's leave this alone for the time being.

Eideren commented:
While we're on the topic of matrix and optimizations, I pulled the simd implementation of dotnet's ``matrix * matrix``, changed it slightly and it works really well as long as we don't use their type; their method signatures forces defensive copies everywhere and our matrices are column major while dotnet's are row major.
```
public static void Matrix4x4Mult<T>(ref T left, ref T right, out T outV) where T : unmanaged
{
    unsafe
    {
        if( sizeof(T) != sizeof(float) * 4 * 4 )
            throw new ArgumentException( $"{typeof(T)}'s size does not match the size of standard 4x4 sp matrix" );
        
        fixed( void* lPtr = &left )
        fixed( void* rPtr = &right )
        fixed( void* outPtr = &outV )
        {
            // Swap left-right if row major
            float* leftPtr = (float*)lPtr;
            float* rightPtr = (float*)rPtr;
            float* outM = (float*)outPtr;
            
            var row = Sse.LoadVector128(rightPtr);
            Sse.Store(outM,
                Sse.Add(Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0x00), Sse.LoadVector128(leftPtr)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0x55), Sse.LoadVector128(leftPtr+4))),
                        Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0xAA), Sse.LoadVector128(leftPtr+8)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0xFF), Sse.LoadVector128(leftPtr+12)))));

            row = Sse.LoadVector128(rightPtr+4);
            Sse.Store(outM+4,
                Sse.Add(Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0x00), Sse.LoadVector128(leftPtr)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0x55), Sse.LoadVector128(leftPtr+4))),
                        Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0xAA), Sse.LoadVector128(leftPtr+8)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0xFF), Sse.LoadVector128(leftPtr+12)))));

            row = Sse.LoadVector128(rightPtr+8);
            Sse.Store(outM+8,
                Sse.Add(Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0x00), Sse.LoadVector128(leftPtr)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0x55), Sse.LoadVector128(leftPtr+4))),
                        Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0xAA), Sse.LoadVector128(leftPtr+8)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0xFF), Sse.LoadVector128(leftPtr+12)))));

            row = Sse.LoadVector128(rightPtr+12);
            Sse.Store(outM+12,
                Sse.Add(Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0x00), Sse.LoadVector128(leftPtr)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0x55), Sse.LoadVector128(leftPtr+4))),
                        Sse.Add(Sse.Multiply(Sse.Shuffle(row, row, 0xAA), Sse.LoadVector128(leftPtr+8)),
                                Sse.Multiply(Sse.Shuffle(row, row, 0xFF), Sse.LoadVector128(leftPtr+12)))));
        }
    }
}
```
BenchmarkDotNet:
```
| Method |      Mean |    Error |   StdDev | Rank |
|------- |----------:|---------:|---------:|-----:|
|   SIMD |  59.09 us | 0.300 us | 0.266 us |    1 |
|  Xenko | 179.11 us | 0.413 us | 0.345 us |    2 |
```
Dotnet's multiply with all of the defensive copies is around 150μ.
The average difference between the output from xenko and simd is ~1/3000000, it's perfectly safe to use.
[And this requires dotnet core 3.0+ though, so we can't use this yet](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.x86.sse.add?view=netcore-3.1).

Eideren commented:
LGTM, If you don't have anything to add we'll merge this

Basewq commented:
All done on my end, nothing else to add. Thanks.

Basewq commented:
Sorry, after looking at the TransformComponent again, I've decided to just avoid the quirk entirely and removed the defensive variable. Probably needs a fresh pair of eyes to check this really is the same logic, but I think it is.
Also reduced the `Matrix.Invert` usage to one line which should also be faster since it's not copying the struct before inverting.

This really is the last change, assuming there isn't anything wrong in the logic.

xen2 commented:
LGTM, thanks!

tebjan commented:
@Basewq thanks for spotting this! this issue was the cause for a bug I was hunting on and off for a few weeks now. when we switched from SharpDX to Xenko math, we had a color transformation method that was using a matrix multiplication chain that had this problem. 

today I was debugging again and saw the multiplication with ref parameters used in a chain and remembered this PR. otherwise, I would probably have searched much longer...
